context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java

```diff
@@ -0,0 +1,287 @@
+ * favors simplicity and low memory allocation during insertion. Although the
+ * asymptotics are good, it is optimized for small sizes like less than 20;
+ * "unbelievably large" would be 100.
+ *
+ * <p>Inspired by popcnt-based compression seen in Ideal Hash Trees, Phil
+ * Bagwell (2000). The rest of the implementation is ignorant of/ignores the
+ * paper.
+ */
+public final class PersistentHashArrayMappedTrie<K,V> {
+  final Node<K,V> root;
+
+  public PersistentHashArrayMappedTrie() {
+    this(null);
+  }
+
+  PersistentHashArrayMappedTrie(Node<K,V> root) {
+    this.root = root;
+  }
+
+  /**
+   * Returns the value with the specified key, or {@code null} if it does not exist.
+   */
+  public V get(K key) {
+    if (root == null) {
+      return null;
+    }
+    return root.get(key, key.hashCode(), 0);
+  }
+
+  /**
+   * Returns a new trie where the key is set to the specified value.
+   */
+  public PersistentHashArrayMappedTrie<K,V> put(K key, V value) {
+    if (root == null) {
+      return new PersistentHashArrayMappedTrie<K,V>(new Leaf<K,V>(key, value));
+    } else {
+      return new PersistentHashArrayMappedTrie<K,V>(root.put(key, value, key.hashCode(), 0));
+    }
+  }
+
+  // Not actually annotated to avoid depending on guava
+  // @VisibleForTesting
+  static final class Leaf<K,V> implements Node<K,V> {
+    final K key;
+    final V value;
+
+    public Leaf(K key, V value) {
+      this.key = key;
+      this.value = value;
+    }
+
+    @Override
+    public V get(K key, int hash, int bitsConsumed) {
+      if (this.key == key) {
+        return value;
+      } else {
+        return null;
+      }
+    }
+
+    @Override
+    public Node<K,V> put(K key, V value, int hash, int bitsConsumed) {
+      int thisHash = this.key.hashCode();
+      if (thisHash != hash) {
+        // Insert
+        return CompressedIndex.combine(
+            new Leaf<K,V>(key, value), hash, this, thisHash, bitsConsumed);
+      } else if (this.key == key) {
+        // Replace
+        return new Leaf<K,V>(key, value);
+      } else {
+        // Hash collision
+        return new CollisionLeaf<K,V>(this.key, this.value, key, value);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Leaf(key=%s value=%s)", key, value);
+    }
+  }
+
+  // Not actually annotated to avoid depending on guava
+  // @VisibleForTesting
+  static final class CollisionLeaf<K,V> implements Node<K,V> {
+    // All keys must have same hash, but not have the same reference
+    private final K[] keys;
+    private final V[] values;
+
+    @SuppressWarnings("unchecked")
+    public CollisionLeaf(K key1, V value1, K key2, V value2) {
+      this((K[]) new Object[] {key1, key2}, (V[]) new Object[] {value1, value2});
+      assert key1 != key2;
+      assert key1.hashCode() == key2.hashCode();
+    }
+
+    // Not actually annotated to avoid depending on guava
+    // @VisibleForTesting
+    CollisionLeaf(K[] keys, V[] values) {
```

Comment:
private? It looks like the usages in tests could use the other constructor instead. (This one does less verification, so was intended to be used rarely.)